rfc/1216-resolve-cloud-tokens.md

```diff
@@ -0,0 +1,845 @@
+on resource creation and passing any required information to downstream processes, 
+such as `cdk8s synth`.
+
+> <sup>*</sup> Independent with respect to usage in Kubernetes resource definitions.
+
+However, this separation into imperative provisioning essentially re-introduces all
+the complexity that IaC aims to solve, and is not desirable.
+
+#### Post Synth Processing
+
+In this approach, cloud infrastructure and Kubernetes resources are defined in the same
+application using the standard IaC tooling. The Kubernetes resources are tightly coupled
+with their required cloud resources. 
+
+```ts
+...
+
+// define the bucket without an explicit name
+new aws.aws_s3.Bucket(stack, 'Bucket');
+
+// pass the bucket name to the CronJob container
+new kplus.CronJob(manifest, 'CronJob', {
+ containers: [{ 
+   image: 'job',
+   envVariables: {
+     BUCKET_NAME: kplus.EnvValue.fromValue(bucket.bucketName),
+   }
+ }]
+});
+
+cdk8sApp.synth();
+
+...
+```
+
+To overcome the problem of unresolved tokens, customers have to post-process the result 
+of `cdk8sApp.synth()` to produce the final deployable manifest. This post synthesis 
+step inspects the manifest for references to cloud resources, interprets them, 
+and performs the necessary lookups.
+
+From an application maintenance perspective, this approach is actually pretty robust.
+The problem is that implementing and maintaining such a post synthesis step is not at all
+trivial. This RFC essentially proposes baking this step into the cdk8s framework, so that
+customers don't have to deal with the complexities it poses.
+
+### What is the technical solution (design) of this feature?
+
+The high-level design consists of three parts:
+
+#### Custom Resolvers
+
+When cdk8s applications are synthesized, the [`resolve`](https://github.com/cdk8s-team/cdk8s-core/blob/v2.7.52/src/_resolve.ts#L3) 
+function is called for every property of the user defined resource spec. Currently, it handles 
+resolving instances of [`Lazy`](https://github.com/cdk8s-team/cdk8s-core/blob/v2.7.52/src/_resolve.ts#L10) 
+and [implicit tokens](https://github.com/cdk8s-team/cdk8s-core/blob/v2.7.52/src/_resolve.ts#L16).
+
+To allow the lookups described in this RFC, we will provide a custom resolution hook in this 
+function in the form of an interface, that customers can implement:
+
+```ts
+import { ApiObject } from './api-object';
+
+export class ResolutionContext {
+
+  public readonly obj: ApiObject;
+  public readonly key: string[];
+
+}
+
+export interface IResolver {
+
+  resolve(context: ResolutionContext, value: any): any;
+}
+```
+
+When a cdk8s chart is defined, a custom resolver can be passed to it:
+
+```ts
+new Chart(app, 'Chart', { resolver: new MyClassThatImplementsIResolver() })
+```
+
+During synthesis, the `resolve` function of the custom resolver will be invoked on every primitive value.
+The return value of this function will replace the original value and be written into the manifest.
```

Comment:
Do we do queries for every value on every synthesis? Isn't that a bit wasteful?

Reply:
Also this breaks network-isolated synthesis -- is there a way we can write instructions somewhere to "query later" ?

Reply:
> Do we do queries for every value on every synthesis? Isn't that a bit wasteful?

Possibly. Are you thinking about introducing a context file? I'm not keen on doing that before we see it actually poses a problem but interested what you think.

> Also this breaks network-isolated synthesis -- is there a way we can write instructions somewhere to "query later" ?

In cdk8s there is only `synth`, no `deploy`. So there is no "later"...yes this means that isolated environment won't be able to synthesize such applications. This does make me think that maybe we shouldn't fail synthesis in case we can't perform the query, because it may also have a bad affect on local devs executing synth - they might not have the necessary credentials, or may not even need that part of the manifest for what they are currently developing an testing.

Reply:
I wonder if it might make sense to make this a 2-step process.

The framework/resolver writes instructions on how to fetch the value to a file.

When can then choose to have:

```
# Do everything
$ cdk8s synth 

# Do it in 2 steps, second one runs with necessary credentials, potentially later
$ cdk8s synth --no-fetch

$ export AWS_ACCESS_KEY_ID=****
$ cdk8s fetch-values
```

Reply:
I would use `--no-resolve` and `cdk8s resolve` as the semantic, because not all resolvers necessarily "fetch" values. Let me think about this.